Modify unittest handling

schveiguy opened the pull request:
Currently, if unit tests run and all pass, the program is started. Typically users expect the main function to exit immediately, so they instrument their main program with

``` D
version(unittest) { void main() {} }
else int main(string[] args)
{
   ...
}
```

But this sucks. D shouldn't be running the main program after unit testing (by default). This can be an opt-in for the user, but should default to exiting.

This modification changes the unit test handler to return `int` instead of `bool` to signify different handling of the unit test result. If the handler function returns 0, then the main function is run. If the handler returns nonzero, the main function is not run. If the result is `int.min`, then the program exits with a failure exit code. If the result is negative, the program prints the negated value as the number of modules that failed, and exits with a failure exit code. If the result is positive, that number is printed as the number of modules that were tested, and exits with a success exit code.

the API for setting the handler routine should be backwards compatible, as I allow setting of the handler routine as a bool return or an int return. The one place where it may be incorrect is if you fetch the handler routine and you set a bool version. This code casts it to an int-returning routine. The effect may be very negligible, as I believe both would return using a register. I did the cast mainly to allow code that checks to see which exact routine is installed will work. I'm expecting there's almost no usage of the getter property there anyway.

Possible improvements -- we may want to only enable this behavior with a `-DRT` switch. Or we may want to allow the previous behavior with such a switch. We may want to separate the bool/int handlers into a different API so the getter makes more sense if complaints arise.

I'm adding this for testing and also to allow for discussion.

schveiguy commented:
ping @wilzbach circleci seems very sensitive to line number changes. Can we fix this? I don't like the idea of having the build break because you added some lines of code!

jacob-carlborg commented:
Not sure I fully follow when the different int values will be used, but can't you use an enum?

schveiguy commented:
> Not sure I fully follow when the different int values will be used

So let's say 5 modules are tested. In the default routine, if all 5 unit test routines pass, then the function returns 5. Then the runtime sees that some unit tests were run, prints "5 Module unit tests passed", and exits.

If 2 of the module unit tests fail, stack traces are printed, then the routine returns -2. The runtime then prints "2 Module unit tests failed", and exits.

I could change to enum, but then I'll just print "All unit tests passed". The extra info of how many modules passed may be trivia, but it may also be useful. I'm open to being convinced 😉

schveiguy commented:
Fixed issue with trace expectation. This will probably fix circleci tests too.

jacob-carlborg commented:
I see. Another thing, how does this work with custom unit test runners? For example, a unit test runner that knows how many actual unit tests were run instead of the number of tested modules. That unit test runner would most likely not want print the number of modules tested.

schveiguy commented:
The customer runner can return 0 or int.min, and nothing will be printed. However, you may want the runtime not to run main, but still return 0 without printing anything. That case isn't handled. Perhaps int.max should do that.

schveiguy commented:
> Perhaps int.max should do that

Now does that. Code looks more even also.

schveiguy commented:
BTW, part of the WIP, I need to add testing for these cases.

jacob-carlborg commented:
The `int.max` addition should be documented as well.

schveiguy commented:
@jacob-carlborg updated docs.

jacob-carlborg commented on src/core/runtime.d:
```diff
@@ -300,6 +301,30 @@ struct Runtime
      * value of this routine indicates to the runtime whether the tests ran
      * without error.
      *
+     * There are two options for handlers. The `bool` version is deprecated but
```
Shouldn't this say "schedule for deprecation" unless `deprecated` is added to the deprecated overload?

schveiguy replied:
This is going to be permanently left in place, in my opinion. It doesn't cause much grief, the translation to the new version is backwards compatible, and so we don't have to do anything more.

We may want to simply undocument it at some point.

jacob-carlborg replied:
Ok, fair enough.

deadalnix commented:
I'm not convinced this is the right approach. The thing will still fail to link when no main function is provided.

IMO, it is better to do this as proposed by basil. I plan to do a DMD PR.

schveiguy commented:
> The thing will still fail to link when no main function is provided.

dmd -main should work just fine.

schveiguy commented:
@wilzbach the issue actually wasn't circleci, it was the build itself (see my later [comment](https://github.com/dlang/druntime/pull/1685#issuecomment-256743553)). I didn't realize this right away, so false alarm!

andralex commented on src/core/runtime.d:
```diff
@@ -300,6 +301,30 @@ struct Runtime
+     *
+     * The default unit tester does not return `int.min` or `int.max`, but
+     * rather the number of tests passed or failed (or 0 if none are run).
+     *
```
OK, so for the new handler:

* 0 means all good run main
* int.min failure with no extra information
* int.max all good don't run main
* >0 these many unittests passed, none failed, don't run main
* <0 the negation of these many unittests failed, print and exit(1)

That's subtle but nothing we can't handle. How about an enum as was suggested already:

How about this:

```d
enum UnittestResult { failStop = int.min, successStop, successContinue = 0 }
```

Then everything outside these values would obey the <0 / >0 rules. For nice symmetry `successStop` could be `int.max`.

schveiguy replied:
I will still make the return be int, but also define an enum that can be returned for convenience.

andralex commented on src/core/runtime.d:
```diff
@@ -328,46 +354,58 @@ struct Runtime
     static @property ModuleUnitTester moduleUnitTester()
     {
+        if(sm_moduleUnitTester == null)
+            return cast(ModuleUnitTester)sm_boolModuleUnitTester;
```
Instead of the cast use a small lambda?

schveiguy replied:
So I'm a bit concerned about doing it that way. I don't know what code exists out there, but there is the possibility that code is checking to see if their particular unit tester is set. If I wrap in a lambda, that will always be false.

Now, code that checks against the function address of wrong type is going to fail to compile. But there will be no way to rectify the situation. At least with the way I have implemented, you can cast to the correct type to figure out if you've set the function.

Alternatively, I can just expose the "new" style unit tester via a new property (and make them mutually exclusive).

schveiguy replied:
Also note, that there isn't any danger of improperly calling the function, since it takes no parameters. Just the return type is different, and won't crash the code that uses the wrong function type.

andralex commented on src/core/runtime.d:
```diff
@@ -492,32 +538,34 @@ extern (C) bool runModuleUnitTests()
+                catch( Throwable e )
+                {
+                    _d_print_throwable(e);
+                    --failed;
```
cool :) prolly clearer if the name were `negativeFailed`

schveiguy replied:
I will fix the code to increment instead (and return `-failed` at the end).

andralex commented on src/rt/dmain2.d:
```diff
@@ -468,8 +468,26 @@ extern (C) int _d_run_main(int argc, char **argv, MainFunc mainFunc)
+            else if (utResult > 0)
+            {
+                if(utResult != int.max)
+                    .fprintf(.stderr, "%d Module unit tests passed\n", utResult);
```
use the actual "unittest" keyword:

```d
.fprintf(.stderr, "%d unittests passed\n", utResult);
```

schveiguy replied:
ok

andralex commented on src/rt/dmain2.d:
```diff
@@ -468,8 +468,26 @@ extern (C) int _d_run_main(int argc, char **argv, MainFunc mainFunc)
+            else
+            {
+                if (utResult != int.min)
+                    .fprintf(.stderr, "%d Module Unit tests failed\n", -utResult);
```
similar here, maybe FAILED in uppercase

schveiguy replied:
ok

schveiguy commented:
I will finish this up with some testing soon!

andralex commented:
So this would be complemented by a separate flag (e.g. '-unittest-only') in the compiler.

CyberShadow commented:
> So this would be complemented by a separate flag (e.g. '-unittest-only') in the compiler.

Would it be possible to make it so that DMD's `-main` and rdmd's `--main`, when combined with `-unittest`, both continue to do the right thing after the change?

andralex commented:
@CyberShadow I was thinking one uses either `-unittest` or -unittest-only` but not both.

schveiguy commented:
Made suggested edits.
Added test harness for custom unittest handler
Added support for `--DRT-unittestmode` option:
* `runmain`: run main even on passing unittests (currently default, change in 2 versions).
* `unittestonly`: if any unittests present, do not run main. (will be default on 2.074)

schveiguy commented:
Note, automated coverage is not going to work here, because of the nature of how we have to test these things. The added tests should cover all the new code however.

jacob-carlborg commented on src/core/runtime.d:
```diff
@@ -492,32 +562,48 @@ extern (C) bool runModuleUnitTests()
+    {
+    case "":
+        // By default, run main. Switch to only doing unit tests in 2.074
+    case "runmain":
```
Could we please call this `run-main`.

schveiguy replied:
Assuming you are backing out of that request?

jacob-carlborg replied:
Yes. I accidentally commented on the wrong line. I left the comment because I hate when I get a notification and then I get find it on GitHub because it was removed.

jacob-carlborg commented on src/core/runtime.d:
```diff
@@ -492,32 +562,48 @@ extern (C) bool runModuleUnitTests()
+    import rt.config : rt_configOption;
+    if (failed != 0)
+        return -failed;
+    else switch (rt_configOption("unittestmode"))
```
Could we please call this `unit-test-mode` or `unittest-mode`.

schveiguy replied:
I'm game for anything in terms of naming, I agree it's long but I couldn't think of something better. My original does follow precedent of all lowercase no-space or hyphen.

CC @MartinNowak thoughts?

jacob-carlborg replied:
I don't think the name is too long, I just like to have a hyphen (or something else) to separate the words. I see three existing usages of `rt_configOption`, two with lowercase no-space or hyphen `gcopt` and `oncycle`  and one with camel case no-space or hyphen `callStructDtorsDuringGC`.

schveiguy replied:
Hm... if there is precedent, I'd prefer the camel case. I was going off [this comment](https://github.com/dlang/druntime/pull/1602#issuecomment-251509511) when I was doing the cycle detection parameters.

jacob-carlborg replied:
I cannot recall ever seeing a program using camel case for the command line flags, most use a hyphen. I don't feel strongly about it, I just thought we should do what most other programs are doing.

wilzbach replied:
I agree with @jacob-carlborg here. Command line flags usually use a hyphen.

jacob-carlborg commented on src/rt/dmain2.d:
```diff
@@ -468,8 +468,26 @@ extern (C) int _d_run_main(int argc, char **argv, MainFunc mainFunc)
+            }
+            else if (utResult > 0)
+            {
+                if(utResult != int.max)
```
* Space after `if`
* How about using the enum instead of `int.max/min`

schveiguy replied:
> Space after if

ok



> How about using the enum

The enum is in the public interface, not imported here. I could re-create the enum, but I don't think it's worth doing that. I will put a comment to clarify that it is attached to the enum.

jacob-carlborg replied:
> The enum is in the public interface, not imported here. I could re-create the enum, but I don't think it's worth doing that. I will put a comment to clarify that it is attached to the enum.

Works for me. I didn't noticed they were in separate modules.

schveiguy commented:
Fixed formatting nit, and added comments to clarify enum usage. Also squashed into 2 commits.

schveiguy commented on test/unittest/Makefile:
```diff
@@ -0,0 +1,44 @@
+	$(QUIET)test $(HASMAIN) -eq 0 || grep -q main $@
+	$(QUIET)test $(HASMAIN) -eq 1 || ! grep -q main $@
+	$(QUIET)test -z "$(TESTTEXT)" || grep -q "unittests $(TESTTEXT)" $@
+	$(QUIET)test -n "$(TESTTEXT)" || ! grep -q "unittests" $@
```
BTW, if some Make guru out there knows of a better way to do these grep acrobatics, I'm so very willing to fix it :)

andralex replied:
seems par for the course :)

schveiguy commented:
Ping

andralex commented:
The return codification seems still arcane and insufficient. E.g. you can figure out how many failed but not the total run so you can't print "5/160 unittests failed". How about using some good old `ref uint` parameters that are filled with the successful and the failed unit tests instead?

cc @MartinNowak for pulling this

schveiguy commented:
> The return codification seems still arcane and insufficient.

I'm not trying to rule the unittest world here. The purpose of the return value change is to allow 3 different states: "pass and exit", "fail and exit", or "pass and continue". This is vs. the existing `bool` return which just allows "pass and continue" or "fail and exit". The addition of the number of tests passed or failed was just a "bonus" to give a bit of extra info since we have to return something more anyway (I also never liked the fact that when unit tests passed, no indication of passing was printed).

Full unittest suites are going to have their own mechanism of printing the results vs. the crude and limited built-in features of druntime.

andralex commented:
Understood. All I'm saying is let's use two integrals, one for passed and one for failed. It's reasonable and reasonably easy to implement. Thanks.

schveiguy commented:
OK, I will try it out and see what happens.

wilzbach commented:
> OK, I will try it out and see what happens.

Btw as no one mentioned it yet, why don't we return a simple `UnittestResult` `struct`?

schveiguy commented:
> Btw as no one mentioned it yet, why don't we return a simple UnittestResult struct?

May as well if we are breaking the API anyway. @andralex any objection?

andralex commented:
@schveiguy I approve

MartinNowak commented:
> wilzbach added this to the 2.074.0 milestone 9 days ago

Please don't add milestones to PRs unless it __has__ to be part of the release.

schveiguy commented:
I have updated the code (boy that took a while for me to get to), and now the API is as follows:

1. Instead of overloading the moduleUnitTester property, I just added an `extModuleUnitTester` property, to avoid issues with casting function pointers.
2. If the new property is set the old is ignored.
3. The new property now returns a struct, which has lots of cool goodies in it, making the behavior configurable quite easily.
4. I added an opCast!bool, so any code that currently checks `if(runModuleUnitTests)` or the opposite, then it should work.
5. I *hope* all the documentation is properly updated. I'll take a look once the doc generator runs.

schveiguy commented:
ping @andralex @jacob-carlborg @CyberShadow

jacob-carlborg commented on src/core/runtime.d:
```diff
@@ -37,9 +37,58 @@ extern(C) int rt_init();
+    /**
+     * Should we print a summary of the results. Ignored if 0 tests executed.
+     */
+    bool summarize;
```
I'm not sure what the summary actually contains but many frameworks print the number of tests run even if 0 tests were run.

schveiguy replied:
OK, so D's determination if we are in "unittest mode" is whether there exists any unit tests at all. So if there are no unit tests run, the runtime thinks that you are not testing.

But thinking about this some more, we can easily add an option for "don't ever run main, even without unit tests" as well, as we are already processing a runtime switch. I'll add it, and let you know.

BTW, the summary is done in the `dmain2` file, I added it in this PR, look for the `printf` statements

jacob-carlborg commented on src/core/runtime.d:
```diff
@@ -440,14 +535,40 @@ extern (C) void profilegc_setlogfilename(string name);
+ * the original behavior of the unit testing system.
+ *
+ * If no unittest custom handlers are registered, the following algorithm is
+ * executed (the behavior can be affected by the `--DRT-unittestmode` switch
```
I'm not sure what the naming conventions are when it comes to `--DRT` flags but I think it should be called `--DRT-unit-test-mode` since that's what most Posix commands use.

jacob-carlborg commented on src/core/runtime.d:
```diff
@@ -440,14 +535,40 @@ extern (C) void profilegc_setlogfilename(string name);
+ *
+ * If the --DRT-unittestmode is set to "runmain", then even if unit tests are
+ * run (and all pass), main is still run. This is currently the default. If
+ * --DRT-unittestmode is set to "unittestonly", then any unit tests present
```
Same here, I think `unittestonly` should be `unit-test-only`.

andralex replied:
well unittest is already considered a special word so prolly unittest-only etc

schveiguy replied:
OK, will do.

schveiguy commented:
I added a third option for the DRT parameter. I changed the parameter to just say "test" and not "unittest", since it was very verbose otherwise. Let me know what you think. I still need to make a changelog entry.

jacob-carlborg commented on src/core/runtime.d:
```diff
@@ -328,45 +390,78 @@ struct Runtime
+     *  The current  module unit tester handler or null if none has been
+     *  set.
+     */
+
```
Please remove empty newline between Ddoc comment and method declaration. It's not consistent with the already existing declarations.

jacob-carlborg commented on src/core/runtime.d:
```diff
@@ -328,45 +390,78 @@ struct Runtime
      * }
      * ---------
      */
+    static @property void extModuleUnitTester( ExtendedModuleUnitTester h )
```
I would prefer that the name, `ext`, was not shortened. Most likely you only going to call this method once, there's no reason to make it less clear to save a few letters.

schveiguy replied:
I wished to call it `moduleUnitTester`, but since there is an accessor, I'm reluctant to do that (the previous version of this PR did that, but the function pointer was returning a value compatible with bool).

I can change this.

jacob-carlborg commented on src/core/runtime.d:
```diff
@@ -440,14 +535,45 @@ extern (C) void profilegc_setlogfilename(string name);
+ * the original behavior of the unit testing system.
+ *
+ * If no unittest custom handlers are registered, the following algorithm is
+ * executed (the behavior can be affected by the `--DRT-testmode` switch
```
I would have hoped for `--DRT-testmode` to be renamed to ``--DRT-test-mode` as well.

schveiguy replied:
So far, the DRT options are gcopt, oncycle, and it looks like some other newer camelCase ones: callStructDtorsDuringGC, scanDataSeg

So I don't know which precedent to follow. ping @MartinNowak @andralex

schveiguy replied:
I'll note none of the precedents have dashes.

jacob-carlborg replied:
I try to follow what I see in most command line applications. In these case I like to look at how Git is doing things. Git contains quite a lot of subcommands and flags. There's also this, which I think is from the Posix standard.

"The arguments that consist of <hyphen-minus> characters and single letters or digits" [1].

[1] http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap12.html

jacob-carlborg replied:
I can't recall that I've event seen command line flags that use camel casing.

jacob-carlborg commented on src/core/runtime.d:
```diff
@@ -494,32 +620,60 @@ extern (C) bool runModuleUnitTests()
+    }
+    else switch (rt_configOption("testmode"))
+    {
+    case "":
```
Shouldn't everything in this block be indented one level?

schveiguy replied:
Hm... I generally don't indent switch case labels. I'm not sure what the style guide says. I see switch statements elsewhere that don't have case labels indented: https://github.com/dlang/druntime/blob/master/src/core/time.d#L352

jacob-carlborg replied:
Yeah, I was surprised now to see how many place in Phobos that don't indent cases in switch statements. I always indent the content of a block.

jacob-carlborg commented on src/rt/dmain2.d:
```diff
@@ -472,8 +482,27 @@ extern (C) int _d_run_main(int argc, char **argv, MainFunc mainFunc)
+        {
+            auto utResult = runModuleUnitTests();
+            assert(utResult.passed <= utResult.executed);
+            if(utResult.passed == utResult.executed)
```
Space after `if` before the opening parenthesis .

jacob-carlborg commented on src/rt/dmain2.d:
```diff
@@ -472,8 +482,27 @@ extern (C) int _d_run_main(int argc, char **argv, MainFunc mainFunc)
+            else
+            {
+                if (utResult.summarize)
+                    .fprintf(.stderr, "%d/%d unittests FAILED\n", cast(int)(utResult.executed - utResult.passed), cast(int)utResult.executed);
```
Most test frameworks will print the same summary regardless if something failed or not. That is, they usually print the total number of tests, the number of passed and the number of failed. I'm not sure how we would like it to behave.

schveiguy replied:
If nothing failed, then the number of tests run and passed is printed (a few lines above this).

This behavior is totally up for grabs if someone wants to make the effort to repaint the bikeshed. My goal is simply to stop the runtime from always executing main after running unit tests.

jacob-carlborg commented:
> I added a third option for the DRT parameter.

👍 

> I changed the parameter to just say "test" and not "unittest", since it was very verbose otherwise. 

I like that, then we don't have to argue if it should be `unit-test` or `unittest` 😃.

jacob-carlborg commented:
Please squash all commits when done.

schveiguy commented on src/core/runtime.d:
```diff
@@ -37,9 +37,58 @@ extern(C) int rt_init();
+     *    true if execution should continue after testing is complete, false if
+     *    not.
+     */
+    bool opCast(T : bool)() const
```
It's not to be cute :)

It's to not break code. Existing code that may have their own runtime startup could easily be written like:

```D
if(runModuleUnitTests())
{
   dmain();
}
```

schveiguy commented on src/core/runtime.d:
```diff
@@ -328,45 +390,78 @@ struct Runtime
+     *     if(result.executed != result.passed)
+     *         result.summarize = true; // print failure
+     *     else
+     *         result.runMain = true; // all UT passed
```
Hm... I didn't set `summarize` or `runMain` if they were to be `false` (I did originally, similarly to what you had, but realized it defaults to `false` anyway). I can be more explicit.

schveiguy replied:
Oops, didn't realize this was the example. The real code is more complex due to the `DRT` options. I've updated it anyway.

andralex commented on src/core/runtime.d:
```diff
@@ -328,45 +390,78 @@ struct Runtime
      *         }
      *     }
-     *     return failed == 0;
+     *     if(result.executed != result.passed)
```
space after if, mmmmkay? :)

schveiguy replied:
You could use torture on me, and I still won't ever remember to do this 😆

schveiguy commented:
OK, I think this is good. Please let me know if I missed any spacing issues!
I added a changelog entry, please review also.
